src/_pytest/mark/__init__.py

```diff
@@ -218,17 +223,29 @@ class MarkMatcher:
     Tries to match on any marker names, attached to the given colitem.
     """
 
-    __slots__ = ("own_mark_names",)
+    __slots__ = ("own_mark_name_mapping",)
 
-    own_mark_names: AbstractSet[str]
+    own_mark_name_mapping: dict[str, list[Mark]]
 
     @classmethod
-    def from_item(cls, item: Item) -> MarkMatcher:
-        mark_names = {mark.name for mark in item.iter_markers()}
-        return cls(mark_names)
+    def from_markers(cls, markers: Iterable[Mark]) -> MarkMatcher:
+        mark_name_mapping = collections.defaultdict(list)
+        for mark in markers:
+            mark_name_mapping[mark.name].append(mark)
+        return cls(mark_name_mapping)
 
-    def __call__(self, name: str) -> bool:
-        return name in self.own_mark_names
+    def __call__(self, name: str, /, **kwargs: str | int | bool | None) -> bool:
+        if not (matches := self.own_mark_name_mapping.get(name, [])):
```

Comment:
This var isn't actually used within the block so there's no need to have walrus here:
```suggestion
        if not self.own_mark_name_mapping.get(name, []):
```

Reply:
Yes it is on L244.

Reply:
Ah, okay, I got confused with how it works.